wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Content/MessageActions/MessageActionsViewController.swift

```diff
@@ -0,0 +1,111 @@
+
+    static func controller(withActions actions: [MessageAction],
+                           actionController: ConversationMessageActionController) -> MessageActionsViewController {
+        let controller = MessageActionsViewController(title: MessageLabelMarker,
+                                            message: nil,
+                                            preferredStyle: .actionSheet)
+        controller.addMessageActions(actions, withActionController: actionController)
+        return controller
+    }
+
+    private var actionController: ConversationMessageActionController?
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+    }
+
+    private func addMessageActions(_ actions: [MessageAction],
+                           withActionController actionController: ConversationMessageActionController) {
+        self.actionController = actionController
+        addReactionsView(withDelegate: self)
+        actions.forEach { addAction($0) }
+        addCancelAction()
+    }
+
+    private func addCancelAction() {
+        let cancelAction = UIAlertAction(title: L10n.Localizable.General.cancel, style: .cancel)
+        addAction(cancelAction)
+    }
+
+    private func addReactionsView(withDelegate delegate: ReactionPickerDelegate) {
+        let reactionPicker = BasicReactionPicker()
+        reactionPicker.delegate = delegate
+        reactionPicker.translatesAutoresizingMaskIntoConstraints = false
+
+        guard let customContentPlaceholder = self.view.findLabel(withText: MessageActionsViewController.MessageLabelMarker),
+              let customContainer =  customContentPlaceholder.superview else { return }
+
+        view.addSubview(reactionPicker)
+        customContentPlaceholder.text = ""
+
+        NSLayoutConstraint.activate([
+            reactionPicker.heightAnchor.constraint(equalToConstant: 72),
+            reactionPicker.widthAnchor.constraint(equalTo: customContainer.widthAnchor),
+            reactionPicker.leadingAnchor.constraint(equalTo: customContainer.leadingAnchor),
+            reactionPicker.topAnchor.constraint(equalTo: customContainer.topAnchor),
+            customContainer.heightAnchor.constraint(equalTo: reactionPicker.heightAnchor)
+        ])
+    }
+
+    private func addAction(_ action: MessageAction) {
+        guard let title = action.title,
+              let actionController = actionController,
+              let selector = action.selector,
+              actionController.canPerformAction(selector)
+        else { return }
+        let style: UIAlertAction.Style = (action == .delete) ? .destructive : .default
+        let newAction = UIAlertAction(title: title, style: style) { [action, weak actionController] _ in
+            actionController?.perform(action: action)
+        }
+        if let image = action.icon?.makeImage(size: .small, color: SemanticColors.Icon.foregroundDefaultBlack) {
+            newAction.setValue(image, forKey: "image")
```

Comment:
Shouldn't be an icon name and not just "image"?

Reply:
I don't understand

Reply:
i don't understand